CI: Enforce Conventional Commits using `commitlint`

0xquark opened the pull request:
<!-- Please read https://rucio.cern.ch/documentation/contributing before submitting a pull request -->
Fixes #7945 

I am using [commitlint](https://commitlint.js.org/) as a linter for our commit messages to enforce conventional commits. It is widely used in the community and [is properly maintained](https://github.com/conventional-changelog/commitlint). It also offers a pre-commit hook to enable it locally. 

We are using this format: `<type>(<scope>): <description>` 

eg. `feat(Core): Add new rule evaluation engine`

I've added some types based on the types used in [standardized conventional commits](https://www.conventionalcommits.org/en/v1.0.0/). 
The scope is based on the available labels in rucio and is taken from what the [rucio's contributing guide](https://rucio.github.io/documentation/contributing/) mentions. 

These are the rules that i am enforcing through this:
1. **Type**: Must be one of the allowed types listed above
2. **Scope**: Must be one of the predefined Rucio components (PascalCase)
3. **Description**: 
   - Should not end with a period
   - Should be concise but descriptive
4. **Issue Number**: Must reference a GitHub issue using [git-trailer](https://alchemists.io/articles/git_trailers) with `#<number>`
     - In order to add a git trailer, you need to commit using `git commit -m "chore(client): fix xyz" --trailer "Closes: #Issue Number"`
     - Only `closes` and `issue` trailer are currently supported. 
     - Note: You can also add a git-trailer by just adding the line at the end of the commit body
     
5. **Line Length**: Header must not exceed 100 characters (This prevents commit titles from being truncated in the GitHub UI. Additional details can always be included in the commit body. )

Attaching screenshot of how it works: 
<img width="1501" height="785" alt="image" src="https://github.com/user-attachments/assets/ad034dbf-a5a1-462c-b490-aba859dd9fe0" />

Note: The only drawback I can foresee is that commitlint is written in JavaScript, while Rucio is a Python-heavy project. To use this tool, we would need to maintain a small JavaScript configuration and install it via npm, which requires maintaining a package.json file.

dchristidis commented on commitlint.config.js:
```diff
@@ -0,0 +1,77 @@
+      return [true];
+    }
+  }
+};
```
This is what we had discussed at the Rucio retreat, but since then I’ve been warming up to the idea of a Git trailer instead. It feels like a more appropriate method to include such information in the commit message.

mgajek-cern commented:
Good work. Given we're adding a new toolchain, would it be worth doing a quick comparison? Maybe something like commitlint vs a simple Python validator vs GitHub Actions-only? Just a basic pros/cons on complexity, features, and maintenance - nothing fancy. It would help us feel confident we're making the right call and give us something to reference later.

0xquark commented:
> Good work. Given we're adding a new toolchain, would it be worth doing a quick comparison? Maybe something like commitlint vs a simple Python validator vs GitHub Actions-only? Just a basic pros/cons on complexity, features, and maintenance - nothing fancy. It would help us feel confident we're making the right call and give us something to reference later.

Thanks @mgajek-cern! I agree a quick comparison would help, but I don’t think it’s worth the extra effort and time. My first thought is to use something that’s already well maintained. Since `commitlint` is widely used and actively maintained, I’d prefer that option. Creating our own GitHub Action or Python validator would just add another script for us to maintain, and we’re already trying to cut down on extra scripts with the test-infra refactoring. Among the commit linters I looked at, commitlint seems more popular in the community and offers better features and long-term support.
@rdimaio @bari12 thoughts?

rdimaio commented:
> > Good work. Given we're adding a new toolchain, would it be worth doing a quick comparison? Maybe something like commitlint vs a simple Python validator vs GitHub Actions-only? Just a basic pros/cons on complexity, features, and maintenance - nothing fancy. It would help us feel confident we're making the right call and give us something to reference later.
> 
> Thanks @mgajek-cern! I agree a quick comparison would help, but I don’t think it’s worth the extra effort and time. My first thought is to use something that’s already well maintained. Since `commitlint` is widely used and actively maintained, I’d prefer that option. Creating our own GitHub Action or Python validator would just add another script for us to maintain, and we’re already trying to cut down on extra scripts with the test-infra refactoring. Among the commit linters I looked at, commitlint seems more popular in the community and offers better features and long-term support. @rdimaio @bari12 thoughts?

commitlint is listed in the suggested tooling for conventional commits https://www.conventionalcommits.org/en/about/#tooling-for-conventional-commits and it apparently seems to be the most popular one as far as GH stars, and it seems to fit our use case without us needing to maintain an extra tool, so I'm inclined to vote for it too.

maany commented on .github/workflows/commitlint.yml:
```diff
@@ -0,0 +1,33 @@
+      - name: Setup Node.js
+        uses: actions/setup-node@49933ea5288caeca8642d1e84afbd3f7d6820020 # v4.4.0
+        with:
+          node-version: '20'
```
Node 20 is in maintenance mode, could we use a more recent version

mgajek-cern commented on commitlint.config.js:
```diff
@@ -0,0 +1,92 @@
+const rucioComponents = [
```
The 24 components will need maintenance as the project evolves. Consider if this list accurately reflects current Rucio architecture.

0xquark replied:
I've taken the list from what's mentioned in the [Rucio Documentation](https://rucio.github.io/documentation/contributing/), which links to the components from [labels](https://github.com/rucio/rucio/labels), and the list is updated with the latest components.

mgajek-cern commented on commitlint.config.js:
```diff
@@ -0,0 +1,92 @@
+  'patch'     // Small fixes or improvements
+];
+
+const gitTrailerPlugin = {
```
Original issue wanted `#<issue_number>` in subject line, but this uses Git trailers instead. Need to align on which approach.

The original issue requested commits like: 

```
fix(Core): Add something #123
```

This implementation expects:

```
fix(Core): Add something

Fixes: #123
```

0xquark replied:
As discussed [here](https://github.com/rucio/rucio/pull/7972#discussion_r2322102298) and during Rucio's Community Meeting, we’re going with the git-trailers approach. I'll update the issue as well.

mgajek-cern commented on .github/workflows/commitlint.yml:
```diff
@@ -0,0 +1,33 @@
+        if: github.event_name == 'push'
+        run: npx commitlint --from HEAD~1 --to HEAD --verbose
+
+      - name: Validate PR commits
+        if: github.event_name == 'pull_request'
+        run: npx commitlint --from ${{ github.event.pull_request.base.sha }} --to ${{ github.event.pull_request.head.sha }} --verbose
```
Should we validate every commit in a PR, or focus on ensuring only the final/merge commit follows the convention? This would allow for informal development commits while maintaining clean project history.
```suggestion
    - name: Validate PR commits
      if: github.event_name == 'pull_request'
      run: npx commitlint --from HEAD~1 --to HEAD --verbose
```

0xquark replied:
I think for the PRs, it makes sense to validate all the commits to ensure consistent history and not only the final commit when merging.

mgajek-cern reviewed (commented):
My few cents.
Solid technical implementation. A few considerations came up during review

dchristidis commented on commitlint.config.js:
```diff
@@ -0,0 +1,92 @@
+  'Clients', 
+  'Database',
+  'Authentication',
+  'Authorisation',
```
This should be one component.

0xquark replied:
Aren’t these two things different in Rucio? Authentication is about validating who the user is, while authorization is about what the user is allowed to do. I do agree that having a single, similar component could help avoid confusion.

dchristidis replied:
Not according to the [list of components](https://rucio.cern.ch/documentation/component_leads).

dchristidis replied:
Maybe it’s better to use the shorthand `Auth`?

dchristidis commented on commitlint.config.js:
```diff
@@ -0,0 +1,92 @@
+  'ci',       // Changes to CI configuration files and scripts
+  'chore',    // Other changes that don't modify src or test files
+  'revert',   // Reverts a previous commit
+  'patch'     // Small fixes or improvements
```
How did you imagine the purpose of `patch`?

0xquark replied:
A `patch` is a small change that isn’t a bug fix (`fix`) or a new feature (`feat`). It’s usually for minor tweaks or adjustments that need tracking but don’t clearly fit other categories. I added it as a fallback for minor changes we’re unsure how to classify, though i feel like it overlaps with other categories like style or refactor.

dchristidis replied:
Could you maybe group with some comments? Maybe these three under ‘functional changes’, most other under ‘non-functional changes’ and the rest under ‘miscellaneous’?

dchristidis replied:
I’m now confused again. `patch` is described as `Small fixes or improvements`. Isn’t that a wildcard for everything? Isn’t there a risk of abuse?

0xquark replied:
I see your point, but I think the idea behind patch was to cover those small tweaks that don’t really count as a bug fix or a feature or a styling improvement. Stuff like adjusting a timeout, changing a small detail in behavior, or improving something minor without adding new functionality/fix.

It’s not meant to be a catch-all, just a way to track small improvements that don’t fit neatly under fix or feat or part of the grey area. As long as we keep the definition clear in our guidelines, I think it should be fine and won’t be abused.
Maybe I'll add a comment or FAQ about not using it as a fallback for everything or explicitly mention it when it is meant to be used.  With Patch, it groups things whose impact is low-risk and incremental (no major functional or structural change).

dchristidis replied:
But even if you’re changing the default timeout, wouldn’t you open a GitHub issue with the `Feature` template?

0xquark replied:
I’d say it depends. If changing the timeout introduces new behavior, it’s a feat; if it’s just tweaking an existing value, it fits patch. That said, I get your point, any change is still an improvement in behavior, no matter how small or big. Now that I see it, I also agree about the potential for patch to be abused, so I’d support removing it from the list.

dchristidis replied:
What type is one supposed to use when removing functionality? I’d say `feat`, but the description doesn’t fit.

I’m also sceptical of `perf`. For one thing, this should probably not be a functional change: you improve performance by doing *the same thing* better. For another, improving performance is `refactor` with a specific goal.

dchristidis replied:
In addition, `feat` and `fix` don’t align very well with the [Contribution Guide](https://rucio.cern.ch/documentation/contributing/#what-should-i-know-before-i-get-started).

bari12 replied:
I understand the worry with `feat` and `fix` - the issue is that these are the only two scopes absolutely required in the conventional commit RFC, all the other scopes are optional. Now, we can say we ignore this, but there is obviously an advantage following the standard. I would propose we keep it like this, for now, and see how it goes. We can still change things later. If I would have to make a guess, the `refactor` scope will be used a lot and is probably the one which we might want to change.

dchristidis replied:
I was coming at it from the opposite direction: `feat` and `fix` should stay, but the distinction between a feature and a patch may need to be adapted. Probably we should pick a name around `BREAKING CHANGE`, as described in the Conventional Commits.

Naming things is hard.

dchristidis replied:
No further comment from me on the proposed types and scopes. We can revise as we gain more experience using them.

dchristidis commented on commitlint.config.js:
```diff
@@ -0,0 +1,92 @@
+      const issueTrailerTokens = [
+        'issue-id', 'issue', 'issues',
+        'closes', 'close', 'closed',
+        'fixes', 'fix', 'fixed',
+        'resolves', 'resolve', 'resolved',
+        'refs', 'ref', 'references', 'reference'
+      ];
```
Is this flexibility desirable? I would prefer to stick to one required trailer (most likely `Issue`), so that I can easily use it in `git log` with a custom format.

0xquark replied:
To me, this seems more about identifying the most useful use case. I’m not entirely sure which direction to take (it’s quite subjective). I personally prioritize developer experience, so I’d like to include these extra options (like `fixes` for closing an issue with my commit, or `refs` for referencing a specific issue). At the same time, you’re also right that it would be simpler to grep for just `Issue`.

dchristidis replied:
I think [PEP 20](https://peps.python.org/pep-0020/#the-zen-of-python) applies here.

> There should be one-- and preferably only one --obvious way to do it.

But I will back off if others have very strong opinions.

0xquark replied:
We talked about this in yesterday’s meeting and agreed on a practical and simple solution: use two trailer tokens `Issue` and `Closes.`

`Issue` will be used to reference an issue.

`Closes` will be used to close an issue through a commit.

Using `Closes` gives us the nice benefit of automatically closing issues when a commit is merged.
And since there are only two trailer tokens, it’ll be easier to grep through them as well.

dchristidis replied:
I remain unconvinced, but won’t push any further.

dchristidis commented on commitlint.config.js:
```diff
@@ -0,0 +1,92 @@
+const rucioComponents = [
```
I would recommend:

 1. Adding a comment before this assignment stating that any changes to this list should also be applied to the GitHub labels and the documentation. A similar comment should probably be added to the latter.
 2. Sorting them alphabetically.
 3. Wherever a shorthand has been applied, adding a comment with the original, e.g. `'Auth', // Authentication & Authorisation`

0xquark commented:
Documentation for Review: https://github.com/rucio/documentation/pull/643

bari12 commented:
To be discussed tomorrow in the dev meeting. But my proposal would be to merge this after the feature freeze next week.

dchristidis commented on commitlint.config.js:
```diff
@@ -0,0 +1,125 @@
+  plugins: [gitTrailerPlugin],
+  rules: {
+    'type-enum': [2, 'always', rucioTypes], // Require a valid type
+    'scope-enum': [2, 'always', rucioComponents], // Require a valid scope
```
What do you do in case you’re making a sweeping change across the entire codebase? For example, we will stop using `typing.Union` once we move to Python 3.12.

0xquark replied:
I get what you mean about a global change not really fitting any specific component. But personally, I think big or breaking changes to the whole codebase are usually easier to deal with if they’re done gradually, with separate commits for each component. It makes the review process a lot more manageable, and if the reviewer wants to squash everything at the end, they still can.

I’m not against adding a global or all scope if we really need it, but I don’t think it’s something we should encourage as the default way to make changes.

dchristidis replied:
That’s probably fine.

Geogouz commented:
Could you rebase please?

0xquark committed:
feat(CI): Enforce conventional commits via commitlint

This commit introduces commitlint to enforce conventional commits in the repository.

Enforced Rules:
- **Types**: `feat`, `fix`, `docs`, `style`, `refactor`, `test`, `ci`, `revert`.
- **Scopes**: Must be one of the defined Rucio components (e.g., `Auth`, `API`, `CI`, `Core`).
- **Issue Trailers**: Commits must include an issue-related Git trailer.
  For example: `Issue: #123` or `Closes: #456`.
- **Breaking Changes**: Commits with breaking changes must indicate this with a `!` in the header
  or a `BREAKING CHANGE` footer, and the two must be consistent.
- **Formatting**:
    - Scope is required.
    - Subject must not end with a period.
    - Body and footer must have a leading blank line.

Changes:
- Added `commitlint.config.js` with Rucio-specific rules.
- Configured allowed types and scopes.
- Implemented custom plugins for issue trailers and breaking change consistency.
- Removed `chore`, `build`, and `perf` from allowed types.
- Added new components like `Auth`.

Issue: #7945

0xquark committed:
feat(CI): disable commitlint for dependabot

Issue: #7945

dchristidis commented on commitlint.config.js:
```diff
@@ -0,0 +1,122 @@
+  'Deletion',
+  'DIRAC',
+  'Documentation',
+  'Lifetime', // Life time model
```
Maybe `LifetimeModel` is preferable here.

dchristidis commented on commitlint.config.js:
```diff
@@ -0,0 +1,122 @@
+      const issueTrailerPattern = new RegExp(`^(${tokenPattern})\\s*[:#]\\s*.+$`, 'im');
+
+      if (!issueTrailerPattern.test(fullMessage)) {
+        return [false, 'Commit message must include an issue-related Git trailer (e.g., "Issue: #123", "Closes: #456"). You can add a trailer using: git commit -m "message" --trailer "Fixes: #123"'];
```
The example should not have ‘Fixes’ is no longer a valid trailer.

0xquark replied:
Nice Catch! 👏

dchristidis commented on commitlint.config.js:
```diff
@@ -0,0 +1,122 @@
+      }
+
+      if (!headerHasBang && hasBreakingFooter) {
+        return [false, 'BREAKING CHANGE footer requires "!" in the commit header (e.g., "chore(api)!: ...").'];
```
Similarly, ’chore’ has been removed and ‘api’ should match case.
